grep/make: remove 'shellpipe', always read from pipe for 'makeef'

aktau opened the pull request:
The commits do not include doc changes yet, as I wanted to get some feedback on whether this is desirable first.

aktau committed:
refactor: replace_makeprg part 2

Mostly allowing some constant strings to actually be declared as such,
which requires plumbing const through a bunch of functions.

Also took the opportunity to replace `char_u` with `char` on
`my_iconv_open` which simplifies some call-sites (thanks to justinmk@
for the comment).

aktau committed:
feat(grep/make): do not write to file if not necessary

Update #6117.

Add a new mode of operation for the Ex commands
`:grep/:lgrep/:make/:lmake`.

If `&shellpipe` and `&makeef` are unset, nvim reads directly from
grepprg's stdout instead of redirecting through a temporary file.
Similar to either of:

    :cgetexpr system(&grepprg ...)
    :silent grep ...

Before this commit, both options being unset resulted in an error:

    :set grepprg=grep\ -n
    :set shellpipe=
    :set makeef=
    :grep restrictions LICENSE
    E40: Can't open errorfile /tmp/nvimD2UTQz/13

The docs describe the behaviour of the old implementation:

> The 'shellpipe' option defaults to ">%s 2>&1" for Win32. This means
> that the output of the compiler is saved in a file and not shown on
> the screen directly.  For Unix "| tee" is used.  The compiler output
> is shown on the screen and saved in a file the same time.  Depending
> on the shell used "|& tee" or "2>&1| tee" is the default, so stderr
> output will be included.
>
> If 'shellpipe' is empty, the {errorfile} part will be omitted.  This
> is useful for compilers that write to an errorfile themselves.

But, the documentation already mentions that this piping may be obiated,
from `:help shellpipe`:

> In the future pipes may be used for filtering and this option will
> become obsolete (at least for Unix).

The new behaviour, if both options are unset:

 - grep/make output is no longer writen to the TTY (see `tee`)
 - Neither is the shell command itself printed.
 - The user also doesn't has to press 'Enter' to return to Neovim.

This mode may obsolete/replace parts of the following customizations:

 - `:help grep` has a ":NewGrep" example.
 - https://gist.github.com/romainl/56f0c28ef953ffc157f36cc495947ab3
 - Plugins that are veneers on top of :grep/:lgrep.
   - https://github.com/george-b/sf.vim

One might think that avoiding (temporary) file I/O is a big boon to
performance, and it might be, but it's not on my (Linux) system:

    $ TEST_FILE=test/functional/ex_cmds/grep_spec.lua  make functionaltest
    -------- Global test environment setup.
    -------- Running tests from test/functional/ex_cmds/grep_spec.lua
    RUN      T1 :grep does not hang on large input #2983: 5731.78 ms OK
    RUN      T2 :grep works when both shellpipe and makeef are unset: 5722.86 ms OK
    -------- 2 tests from test/functional/ex_cmds/grep_spec.lua (11498.24 ms total)

Both tests take ~5 seconds to run. The individual grep commands produce
a decent amount of output, but take a negligible time to run:

    $ cd test/ && time grep a **/* >/dev/null
    ...
    grep --color=auto a **/* > /dev/null  0.01s user 0.01s system 96% cpu 0.018 total

I've profiled this and it's due to errorformat interpretation [2]

[1]: I bet it was `errorformat` parsing. I tried to do
     `feed_command([[set errorformat=%f:%l:%m]])` since the default
     errorformat seems very verbose (and each comma-separate pattern is
     a separate regex). That made no difference to the runtime. It may
     still be `errorformat` parsing, but not as straightforward.

[2]:

perf report --percent-limit 10 --call-graph=none | cat - | head -60
    95.24%     0.00%  nvim      [.] __libc_start_main
    95.24%     0.00%  nvim      [.] main
    95.24%     0.00%  nvim      [.] 0x5541d68949564100
    95.22%     0.00%  nvim      [.] normal_enter
    95.22%     0.00%  nvim      [.] state_enter
    95.04%     0.00%  nvim      [.] normal_execute
    89.37%     0.00%  nvim      [.] do_cmdline
    89.37%     0.00%  nvim      [.] nv_colon
    89.36%     0.00%  nvim      [.] do_one_cmd
    89.36%     0.00%  nvim      [.] execute_cmd0
    89.36%     0.00%  nvim      [.] ex_make
    89.22%     0.01%  nvim      [.] qf_init_ext
    89.20%     0.03%  nvim      [.] qf_init_process_nextline
    88.23%     0.12%  nvim      [.] qf_parse_line
    88.01%     0.26%  nvim      [.] qf_parse_get_fields
    87.71%     0.01%  nvim      [.] vim_regexec
    87.69%     0.13%  nvim      [.] vim_regexec_string
    87.57%     0.02%  nvim      [.] nfa_regexec_nl
    87.54%     2.36%  nvim      [.] nfa_regexec_both
    85.16%     0.10%  nvim      [.] nfa_regtry
    82.96%    23.88%  nvim      [.] nfa_regmatch
    55.78%    41.71%  nvim      [.] addstate

aktau commented:
I'm unsure what's up with the commit linter, it seems to just crash, but the git commit it last prints out runs fine from the shell.

justinmk commented on src/nvim/mbyte.c:
```diff
@@ -2261,7 +2261,7 @@ char_u *enc_locale(void)
 // Returns (void *)-1 if failed.
 // (should return iconv_t, but that causes problems with prototypes).
-void *my_iconv_open(char_u *to, char_u *from)
+void *my_iconv_open(const char_u *to, const char_u *from)
```
could be changed to `char`, all callsites are casting to `char_u` for no reason

aktau replied:
You're right, I should've checked the callsites. I didn't do it because I know theres `char_u` to `char` changes going on in the repo, but in this case it's certainly a benefit. Fixed.

justinmk commented on test/functional/ex_cmds/grep_spec.lua:
```diff
@@ -18,4 +18,20 @@ describe(':grep', function()
+      return
+    end
+
+    feed_command([[set grepprg=grep\ -r]])
```
if command() works that is preferred instead. Not sure why feed_command was used in the existing test.

justinmk commented on test/functional/ex_cmds/grep_spec.lua:
```diff
@@ -18,4 +18,20 @@ describe(':grep', function()
+    feed_command('cd test')
+    feed_command('grep a **/*')
+    feed('<cr>')  -- Press ENTER
+    ok(eval('len(getqflist())') > 9000)  -- IT'S OVER 9000!!1
```
this test is not for UI performance so we could use a more restricted `grep` pattern above. So this test will take less time, and we can assert something more explicit.

aktau replied:
The description of the test includes "large output" though. Can you tell what the minimum thing is we should test? Regardless, I'd prefer fixing that up in a separate commit.

justinmk commented on src/nvim/quickfix.c:
```diff
@@ -4278,36 +4276,73 @@ void ex_make(exarg_T *eap)
-    qi = GET_LOC_LIST(wp);
-    if (qi == NULL) {
+    xfree(out);
+  } else {
```
```suggestion
  } else {  // Read from a temporary file.
```

aktau replied:
Done, also shortened the comment above to be less verbose.

justinmk commented on src/nvim/quickfix.c:
```diff
@@ -4278,36 +4276,73 @@ void ex_make(exarg_T *eap)
+    do_shell(cmd, 0);
+
+    res = qf_init_ext(qi, qi->qf_curlist, fname, curbuf, NULL, errorformat,
+                      newlist, (linenr_T)0, (linenr_T)0, qf_cmdtitle(*eap->cmdlinep), enc);
```
The usual reason for vim's use of temp files with shell commands is to detect the encoding of streamed input. But here the `enc` is being decided beforehand, so what purpose does the temp file serve?

Perhaps in a followup PR or separate commit we could remove this branch entirely?

aktau committed:
test(grep): replace feed_command with command

Proposed in a review comment on PR #20890. The tests seem to work
before/after.

aktau commented:
> 
The scope is checked against a blocklist, and either the : or the / confuses the script. I'd drop the colons and see if that makes it better.

Thanks! That seems to have done it.

@justinmk you left a comment that I see in my mail client but not in this UI, I'll quote and reply here:

> The usual reason for vim's use of temp files with shell commands is to detect the encoding of streamed input. But here the enc is being decided beforehand, so what purpose does the temp file serve?
>
> Perhaps in a followup PR or separate commit we could remove this branch entirely?

I'd love to just simplify this to the common sense thing, but I wasn't sure about the edge cases. Two thoughts:

 1. Reading the documentation it seems that it should work for programs that always output to some local file. That means we could simplify the code a bit more (we could use `os_system` for everything), but we wouldn't be able to get rid of reading from a file (if `makeef` is set).
 2. How would this work? I assume we won't change the `shellpipe` default, so should `:grep/:make` ignore `shellpipe`? That would change the behaviour of :grep/:make to no longer print output to the tty (which I think is good, but it's something to consider). Ignoring `shellpipe` is implied in the current docs, so perhaps we can try that in one go.

I've updated the docs now and force-pushed. PTAL.

aktau commented:
Ugh, I didn't see that the doc files have lots of whitespaces at the end of line, which I have an autocmd for to remove. Now the diff is way too large. I'll try to undo that.

clason commented:
Please don't make unrelated changes to the docs; that's not necessary and will just make porting changes from Vim harder.

aktau commented:
After considering what's written in the 'options.txt' documentation a bit, I'm more willing to change the default behaviour by ignoring `shellpipe` for `:grep/:make/...`. However, this would alter the default behaviour on UNIX that the output is also printed to the TTY. If this is OK with you (and to be fair it was "announced" in the docs), I'll implement it immediately.

> 
Please don't make unrelated changes to the docs; that's not necessary and will just make porting changes from Vim harder. (Yes, that includes whitespace, unfortunately.)

Fine, I'll revert it. Would be nice if someone is willing to strip whitespaces in upstream, because some docs apparently have had this stripped.

aktau commented:
Ping. Should we try the current version or go for "ignore  &shellpipe" immediately?

aktau commented:
Thinking further upon it: `&shellpipe` is only used in `ex_make` AFAICS. So ignoring `&shellpipe` basically means deprecating `&shellpipe`. Which sounds nice. We could go as far as really removing the option.

aktau commented:
Ping. Do you have any opinion on whether I should deprecate `&shellpipe` at once?

aktau committed:
feat(grep/make): read from stdout by default (remove shellpipe)

The 'shellpipe' option is not (that) necessary: read directly from stdout.
This simplifies some of the complexity introduced by an earlier commit
which special-cased an empty 'shellpipe' as meaning: no temporary files.
After this commit :make/:grep will never make temporary files themselves
anymore. The ability to source quickfix/loclist items from a file
remains: if 'makeef' is set, that will be used as the filename to read.

aktau commented:
I've added a new commit that implements the suggestion of removing shellpipe entirely. It does simplify things somewhat. It also modifies a lot of files since shellpipe was referenced in a lot of places. PTAL.

aktau commented:
Ping @justinmk is `feat(grep/make): read from stdout by default (remove shellpipe)` a direction we want to move in? If so, I could squash with the earlier implementation and see if I can fix the tests.

One very noticeable side effect of this is that setting `&shellpipe` after this commit results in an error. I don't think many people set this explicitly, but some scripts do (see all the references I removed from the runtime scripts). It may be worth it to simplify nvim's surface area, but it's the team's call.

aktau committed:
fix(grep/make): also apply 'makeenc' if errorfile is not a file

We're removing 'shellpipe', reading the output of makeprg/grepprg
directly from pipes. The output will no longer be read from temporary
files, but be stored in a vim var instead. We should also apply
requested conversion (enc != menc) to non-file sources (luckily there
were oldtests that checked for this, and failed when I removed
'shellpipe').

    Failures:
    	From test_makeencoding.vim:
    	Found errors in Test_grep():
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_grep line 9: Expected 'Xfoobar.c|10| こ��ん��に��ち��は�� (cp932)' but got 'Xfoobar.c|10| ����ɂ���� (cp932)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_grep line 9: Expected 'Xfoobar.c|10| 你��好�� (cp936)' but got 'Xfoobar.c|10| ��� (cp936)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_grep line 9: Expected 'Xfoobar.c|10| À�È�Ì�Ò�Ù� (latin1)' but got 'Xfoobar.c|10| ����� (latin1)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_grep line 19: Expected 'Xfoobar.c|10| こ��ん��に��ち��は�� (cp932)' but got 'Xfoobar.c|10| ����ɂ���� (cp932)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_grep line 19: Expected 'Xfoobar.c|10| 你��好�� (cp936)' but got 'Xfoobar.c|10| ��� (cp936)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_grep line 19: Expected 'Xfoobar.c|10| À�È�Ì�Ò�Ù� (latin1)' but got 'Xfoobar.c|10| ����� (latin1)'
    	Found errors in Test_make():
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_make line 9: Expected 'Xfoobar.c|10| こ��ん��に��ち��は�� (cp932)' but got 'Xfoobar.c|10| ����ɂ���� (cp932)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_make line 9: Expected 'Xfoobar.c|10| 你��好�� (cp936)' but got 'Xfoobar.c|10| ��� (cp936)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_make line 9: Expected 'Xfoobar.c|10| À�È�Ì�Ò�Ù� (latin1)' but got 'Xfoobar.c|10| ����� (latin1)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_make line 19: Expected 'Xfoobar.c|10| こ��ん��に��ち��は�� (cp932)' but got 'Xfoobar.c|10| ����ɂ���� (cp932)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_make line 19: Expected 'Xfoobar.c|10| 你��好�� (cp936)' but got 'Xfoobar.c|10| ��� (cp936)'
    	command line..script /home/runner/work/neovim/neovim/src/nvim/testdir/runtest.vim[437]..function RunTheTest[44]..Test_make line 19: Expected 'Xfoobar.c|10| À�È�Ì�Ò�Ù� (latin1)' but got 'Xfoobar.c|10| ����� (latin1)'

aktau committed:
test(make): make non-unary-argument count test more robust

In #20890, 'shellpipe' is removed and with it the echoing of the command
line to the window. Thus `{ output = true }` no longer suffices. This
commit adapts the test to be a bit more end-to-end (did the correct line
get matched?).

justinmk commented on runtime/doc/options.txt:
```diff
@@ -5415,39 +5414,6 @@ A jump table for the options with a short description can be found at |Q_op|.
-	security reasons.
-
-						*'shellpipe'* *'sp'*
-'shellpipe' 'sp'	string	(default ">", ">%s 2>&1", "| tee", "|& tee" or
```
Related: https://github.com/neovim/neovim/pull/16271

aktau replied:
Yes, as it turns out powershell was the reason for one of the failing tests (really great that that test exists, even if it was a little too brittle due to parsing output).

justinmk commented on runtime/doc/options.txt:
```diff
@@ -5415,39 +5414,6 @@ A jump table for the options with a short description can be found at |Q_op|.
-	and the other initializations, so that when the 'shell' option is set
-	there, the 'shellpipe' option changes automatically, unless it was
-	explicitly set before.
-	When 'shellpipe' is set to an empty string, no redirection of the
-	":make" output will be done.  This is useful if you use a 'makeprg'
-	that writes to 'makeef' by itself.  If you want no piping, but do
```
This is the key passage.

aktau replied:
The key passage for what? For one of the use cases of `shellpipe`? In that case: this use case is still possible: if `makeef` is set, Neovim will read from that file.

justinmk commented on src/nvim/option.c:
```diff
@@ -621,37 +621,25 @@ void set_init_3(void)
+    p = xstrnsave(p, len);
     //
-    // Default for p_sp is "| tee", for p_srr is ">".
+    // Default for for p_srr is ">".
```
typo

aktau replied:
Thanks, fixed.

justinmk reviewed (approved):
let's try it. you've been looking closely at this, so hoping you will notice any regressions :) 

i'll add a `:h nvim-removed` note in my "docs" pr

justinmk commented:
@aktau there are some failing tests. if those can be fixed, feel free to merge if you don't get a reply after a few days. sorry for the delay!

jamessan commented:
> One very noticeable side effect of this is that setting `&shellpipe` after this commit results in an error. I don't think many people set this explicitly, but some scripts do (see all the references I removed from the runtime scripts).

This will increase the chance of conflicts in the runtime files.  Isn't there existing precedent for options we've removed, but we don't make it an error for the option to be set?  If so, can we follow that pattern here to avoid changing all the runtime scripts?

lewis6991 commented:
This change currently only changes 3 runtime files. Surely that shouldn't cause many conflicts from runtime updates?

aktau commented:
One of the test failures is in make_spec, due to powershell: ttps://github.com/neovim/neovim/blob/08d5b4275ecaae4d3329d4ed2469782f10dd463e/test/functional/ex_cmds/make_spec.lua#L20-L31

The exit code I observe with my new code is `1`, not `3`. I verified that the difference is due to what the `set_shell_powershell` helper does: https://github.com/neovim/neovim/blob/08d5b4275ecaae4d3329d4ed2469782f10dd463e/test/functional/helpers.lua#L587

If this is appended to the `pwsh` command, the exit code of the spawned script is propagated upwards to Neovim. So, for this specific test it seems we "need" (or perhaps: abuse) the shellpipe option to make powershell return the expected value. But it also sort-of seems like an abuse of the shellpipe option.

I'd welcome suggestions of a clean solution to this for this. In the case of the failing test we could just append shellredir, as it is set to the same value, but this will likely bring trouble with more normal shells, and we wanted to get rid of redirection, not have more. cc @3N4N who appears to be the local expert on all things powershell.

3N4N commented:
> I'd welcome suggestions of a clean solution to this for this. In the case of the failing test we could just append shellredir, as it is set to the same value, but this will likely bring trouble with more normal shells, and we wanted to get rid of redirection, not have more. cc @3N4N who appears to be the local expert on all things powershell.

I don't know what 'shellpipe' does.  The little work I have done was restricted to 'shellredir' specifically, so I can't help here.

> Should we perhaps treat powershell specially in `ex_make` as we seem to do in `make_filter_cmd`?

But I can assure, if it helps, that treating powershell specially wouldn't be idiomatically wrong.  PowerShell just doesn't behave like other shells.  So even though generalization is obviously better, specialization for PowerShell shouldn't bring as much concern as for other shells.

aktau committed:
fix(grep/make): append `; exit $LastExitCode` to powershell commands

To make the powershell make_spec test pass after `&shellpipe` was
removed:

1. Check the value of `v:shell_error` instead of parsing `makeprg`
   output. This is more robust in general, and due to the removal of
   `shellpipe` this wasn't present in the output anymore.
2. Before removing `&shellpipe`, it was being (ab)used to force
   powershell to return the exit code of the last-ran command. It does
   not do this by default:
   https://github.com/PowerShell/PowerShell/issues/13501. Without
   `&shellpipe`, we're forced to special-case powershell inside native
   code and append it ourselves.

aktau committed:
fold into shellpipe removal: better handling of empty output

aktau committed:
oldtests: remove reliance on temp files

That we can read tempfiles (`makeef`) when grepprg writes into them is
tested in the grep_spec.lua functional test.

aktau committed:
fix(grep,make): also use buffer-local errorformat when sourcing from typval

This is what made the tests in test_compiler.vim fail. I spent way too
long tracking this down. To think I actually added a TODO to investigate
this strange exclusion before...

aktau committed:
temp commit to track down windows test error

aktau committed:
more windows hacks
